### DISCUSSION TITLE
Make modules failed: CONFIG_RETPOLINE=y, but not supported by the compiler

### DISCUSSION THREAD

**Discussion by wangw42** [NVIDIA/open-gpu-kernel-modules#407] [Q&A]
I’am trying to make modules, but failed.
And I’ve tried to update gcc version, but also failed.
Maybe I should update my linux kernel to solve this problem?

>make[1]: Entering directory `/usr/src/kernels/3.10.0-1160.42.2.el7.x86_64'
/usr/src/kernels/3.10.0-1160.42.2.el7.x86_64/arch/x86/Makefile:96: stack-protector enabled but compiler support broken
/usr/src/kernels/3.10.0-1160.42.2.el7.x86_64/arch/x86/Makefile:166: *** CONFIG_RETPOLINE=y, but not supported by the compiler. Compiler update recommended..  Stop.
make[1]: *** [sub-make] Error 2
make[1]: Leaving directory `/usr/src/kernels/3.10.0-1160.42.2.el7.x86_64'
make: *** [modules] Error 2

BTW, CentOS 7.9 and gcc version 4.8.5 20150623 (Red Hat 4.8.5-44) 

Thanks for your help.

**Comment by danielgora** [comment]
The problem is usually that you are building as a non-root user and the kernel makefiles cannot create a temporary directory in /usr/src/kernels/*version* to be able to detect the cc flags.

When the kernel tries to determine if a CC flag is available it runs the 'try-run' macro in scripts/Makefile.compiler (or scripts/Kbuild.include depending on the kernel version) which creates a temp directory:

```
 # output directory for tests below
 TMPOUT = $(if $(KBUILD_EXTMOD),$(firstword $(KBUILD_EXTMOD))/).tmp_$$$$

 # try-run
 # Usage: option = $(call try-run, $(CC)...-o "$$TMP",option-ok,otherwise)
 # Exit code chooses option. "$$TMP" serves as a temporary file and is
 # automatically cleaned up.
 try-run = $(shell set -e;		\
	TMP=$(TMPOUT)/tmp;		\
	trap "rm -rf $(TMPOUT)" EXIT;	\
	mkdir -p $(TMPOUT);		\
	if ($(1)) >/dev/null 2>&1;	\
	then echo "$(2)";		\
	else echo "$(3)";		\
	fi)

 __cc-option = $(call try-run,\
	$(1) -Werror $(2) $(3) -c -x c /dev/null -o "$$TMP",$(3),$(4))
```

So you can see here that if KBUILD_EXTMOD is not set then $TMPOUT will end up using /usr/src/kernels/*version*.  If you don't have write permissions there, the `mkdir -p $(TMPOUT)` will fail and then the cc-option macro will return nothing.

So the solution is one of the following:
!) build as root.  root has write permission in /usr/src/kernels/*version*
2) `make KBUILD_EXTMOD=$(pwd)`.  This way KBUILD_EXTMOD will be set and $TMPOUT will use the directory where you are building and presumably have write permission.
3) `sudo chmod o+w /usr/src/kernels/<version>/.` .  Just allow other users to write to /usr/src/kernels/<version> so that the macros will work for non-root users.

Note that if you are building a module with something like `make -C /usr/src/kernels/<version> M=$(pwd) modules`, then you shouldn't have to do this because the M flag will set KBUILD_EXTMOD.
